shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -407,123 +244,157 @@ class JSBackend {
+        methodScope.declareValue("this")
+        JSClassMethod(name, methodParams, L(translateTerm(body)(methodScope)))
+      case term =>
+        val getterScope = scope.derive(s"Getter $name")
+        getterScope.declareValue("this")
+        JSClassGetter(name, L(translateTerm(term)(getterScope)))
     }
   }
-}
 
-class JSWebBackend extends JSBackend {
-  // This will be changed during code generation.
-  private var hasWithConstruct = false
+  /**
+    * Declare symbols for types, traits and classes.
+    * Call this before the code generation.
+    * 
+    * @return defined class symbols
+    */
+  protected def declareTypeDefs(typeDefs: Ls[TypeDef]): Ls[ClassSymbol] = typeDefs flatMap {
+    case TypeDef(Als, TypeName(name), tparams, body, _, _) =>
+      topLevelScope.declareTypeAlias(name, tparams map { _.name }, body)
+      N
+    case TypeDef(Trt, TypeName(name), tparams, body, _, _) =>
+      topLevelScope.declareTrait(name, tparams map { _.name }, body)
+      N
+    case TypeDef(Cls, TypeName(name), tparams, baseType, _, members) =>
+      S(topLevelScope.declareClass(name, tparams map { _.name }, baseType, members))
+  }
+
+  /**
+    * Find the base class for a specific class.
+    */
+  private def resolveBaseClass(ty: Type): Opt[ClassSymbol] = {
+    val baseClasses = ty.collectTypeNames.flatMap { name =>
+      topLevelScope.getType(name) match {
+        case S(sym: ClassSymbol) => S(sym)
+        case S(sym: TraitSymbol) => N // TODO: inherit from traits
+        case S(sym: TypeAliasSymbol) =>
+          throw new CodeGenError(s"cannot inherit from type alias $name" )
+        case N =>
+          throw new CodeGenError(s"undeclared type name $name when resolving base classes")
+      }
+    }
+    if (baseClasses.length > 1)
+      throw CodeGenError(
+        s"cannot have ${baseClasses.length} base classes: " +
+        baseClasses.map { _.lexicalName }.mkString(", ")
+      )
+    else
+      baseClasses.headOption
+  }
 
-  // Name of the helper function for `with` construction.
-  private val withConstructFnName = topLevelScope allocateJavaScriptName "withConstruct"
+  /**
+    * Resolve inheritance of all declared classes.
+    * 
+    * @return sorted class symbols with their base classes
+    */
+  protected def sortClassSymbols(classSymbols: Ls[ClassSymbol]): Iterable[(ClassSymbol, Opt[ClassSymbol])] = {
+    // Cache base classes for class symbols.
+    // Note: it may include previously defined class symbols.
+    val baseClasses = Map.from(classSymbols.iterator.flatMap { derivedClass =>
+      resolveBaseClass(derivedClass.actualType).map(derivedClass -> _)
+    })
+    val sorted = try topologicalSort(baseClasses, classSymbols) catch {
+      case e: CyclicGraphError => throw CodeGenError("cyclic inheritance detected")
+    }
+    // We only need class symbols declared in current translation unit.
+    sorted.flatMap(sym => if (classSymbols.contains(sym)) S(sym -> baseClasses.get(sym)) else N)
```

Comment:
Here we resolve the base class for each newly-defined classes and put them in a `Map`. The information is used immediately in `topologicalSort`. Note that `sortClassSymbols` returns all newly-defined classes with their base classes (`Iterable[ClassSymbol, Opt[ClassSymbol]`).